bug: Crash: Sprout note commitment tree must exist if there is a finalized tip

### What happened?

 I expected to see this happen: zebra starts

Instead, this happened: Zebra is in restart crash loop (it is running under systemd). 

0|zebrad  | The application panicked (crashed).
0|zebrad  | Message:  Sprout note commitment tree must exist if there is a finalized tip
0|zebrad  | Location: zebra-state/src/service/finalized_state/zebra_db/shielded.rs:119
0|zebrad  | Metadata:
0|zebrad  | version: 1.4.0+14.gc3b11e3
0|zebrad  | Zcash network: Mainnet
0|zebrad  | running state version: 25.3.0
0|zebrad  | initial disk state version: 25.2.2
0|zebrad  | features: default,getblocktemplate_rpcs,howudoin,indicatif,progress_bar,release_max_level_info
0|zebrad  | branch: main
0|zebrad  | git commit: c3b11e3
0|zebrad  | commit timestamp: 2023-11-16T21:22:34.000000000Z
0|zebrad  | target triple: x86_64-unknown-linux-gnu
0|zebrad  | rust compiler: 1.72.0
0|zebrad  | rust release date: 2023-08-23
0|zebrad  | optimization level: 3
0|zebrad  | debug checks: false


### What were you doing when the issue happened?

_No response_

### Zebra logs

_No response_

### Zebra Version

_No response_

### Which operating systems does the issue happen on?

- [X] Linux
- [ ] macOS
- [ ] Windows
- [ ] Other OS

### OS details

_No response_

### Additional information

_No response_

It doesn't build with rust stable.

```
error[E0658]: use of unstable library feature 'int_roundings'
   --> zebra-network/src/address_book.rs:276:41
    |
276 |         let address_limit = peers.len().div_ceil(ADDR_RESPONSE_LIMIT_DENOMINATOR);
    |                                         ^^^^^^^^
    |
    = note: see issue #88581 <https://github.com/rust-lang/rust/issues/88581> for more information
```

--- a/zebra-state/src/service/finalized_state/disk_format/upgrade/fix_tree_key_type.rs
+++ b/zebra-state/src/service/finalized_state/disk_format/upgrade/fix_tree_key_type.rs
@@ -28,11 +28,6 @@ pub fn run(
     // Writing the trees back to the database automatically updates their format.
     let mut batch = DiskWriteBatch::new();
 
-    // Delete the previous `Height` tip key format, which is now a duplicate.
-    // It's ok to do a full delete, because the trees are restored before the batch is written.
-    batch.delete_range_sprout_tree(upgrade_db, &Height(0), &MAX_ON_DISK_HEIGHT);
-    batch.delete_range_history_tree(upgrade_db, &Height(0), &MAX_ON_DISK_HEIGHT);
-
     // Update the sprout tip key format in the database.
     batch.update_sprout_tree(upgrade_db, &sprout_tip_tree);
     batch.update_history_tree(upgrade_db, &history_tip_tree);
@@ -45,6 +40,24 @@ pub fn run(
     upgrade_db
         .write_batch(batch)
         .expect("updating tree key formats should always succeed");
+
+    // These deletes can be slow due to tombstones for previously deleted keys,
+    // so we do it in a separate batch to avoid data races with syncing (#7961).
+    let mut batch = DiskWriteBatch::new();
+
+    // Delete the previous `Height` tip key format, which is now a duplicate.
+    // This doesn't delete the new `()` key format, because it serializes to an empty array.
+    batch.delete_range_sprout_tree(upgrade_db, &Height(0), &MAX_ON_DISK_HEIGHT);
+    batch.delete_range_history_tree(upgrade_db, &Height(0), &MAX_ON_DISK_HEIGHT);
+
+    // Return before we write if the upgrade is cancelled.
+    if !matches!(cancel_receiver.try_recv(), Err(mpsc::TryRecvError::Empty)) {
+        return Err(CancelFormatChange);
+    }
+
+    upgrade_db
+        .write_batch(batch)
+        .expect("cleaning up old tree key formats should always succeed");
 
     Ok(())
 }

--- a/zebra-state/src/service/finalized_state/disk_format/upgrade/fix_tree_key_type.rs
+++ b/zebra-state/src/service/finalized_state/disk_format/upgrade/fix_tree_key_type.rs
@@ -41,7 +41,7 @@ pub fn run(
         .write_batch(batch)
         .expect("updating tree key formats should always succeed");
 
-    // These deletes can be slow due to tombstones for previously deleted keys,
+    // The deletes below can be slow due to tombstones for previously deleted keys,
     // so we do it in a separate batch to avoid data races with syncing (#7961).
     let mut batch = DiskWriteBatch::new();
 

--- a/zebra-state/src/service/finalized_state/zebra_db/chain.rs
+++ b/zebra-state/src/service/finalized_state/zebra_db/chain.rs
@@ -62,7 +62,7 @@ impl ZebraDb {
             // In Zebra 1.4.0 and later, we only update the history tip tree when it has changed (for every block after heartwood).
             // But we write with a `()` key, not a height key.
             // So we need to look for the most recent update height if the `()` key has never been written.
-            history_tree = self.db.zs_last_key_value(&history_tree_cf);
+            history_tree = self.db.zs_last_key_value(&history_tree_cf).map(|(_height_key, tree_value)| tree_value);
         }
 
         history_tree.unwrap_or_default()

--- a/zebra-state/src/service/finalized_state/zebra_db/chain.rs
+++ b/zebra-state/src/service/finalized_state/zebra_db/chain.rs
@@ -62,7 +62,10 @@ impl ZebraDb {
             // In Zebra 1.4.0 and later, we only update the history tip tree when it has changed (for every block after heartwood).
             // But we write with a `()` key, not a height key.
             // So we need to look for the most recent update height if the `()` key has never been written.
-            history_tree = self.db.zs_last_key_value(&history_tree_cf).map(|(_height_key, tree_value)| tree_value);
+            history_tree = self
+                .db
+                .zs_last_key_value(&history_tree_cf)
+                .map(|(_key: Height, tree_value)| tree_value);
         }
 
         history_tree.unwrap_or_default()

--- a/zebra-state/src/service/finalized_state/zebra_db/chain.rs
+++ b/zebra-state/src/service/finalized_state/zebra_db/chain.rs
@@ -65,7 +65,7 @@ impl ZebraDb {
             history_tree = self
                 .db
                 .zs_last_key_value(&history_tree_cf)
-                .map(|(_key: Height, tree_value)| tree_value);
+                .map(|(_key, tree_value): (Height, _)| tree_value);
         }
 
         history_tree.unwrap_or_default()

--- a/zebra-state/src/service/finalized_state/zebra_db/chain.rs
+++ b/zebra-state/src/service/finalized_state/zebra_db/chain.rs
@@ -59,11 +59,10 @@ impl ZebraDb {
         let mut history_tree: Option<Arc<HistoryTree>> = self.db.zs_get(&history_tree_cf, &());
 
         if history_tree.is_none() {
-            let tip_height = self
-                .finalized_tip_height()
-                .expect("just checked for an empty database");
-
-            history_tree = self.db.zs_get(&history_tree_cf, &tip_height);
+            // In Zebra 1.4.0 and later, we only update the history tip tree when it has changed (for every block after heartwood).
+            // But we write with a `()` key, not a height key.
+            // So we need to look for the most recent update height if the `()` key has never been written.
+            history_tree = self.db.zs_last_key_value(&history_tree_cf);
         }
 
         history_tree.unwrap_or_default()

--- a/zebra-state/src/service/finalized_state/zebra_db/shielded.rs
+++ b/zebra-state/src/service/finalized_state/zebra_db/shielded.rs
@@ -112,7 +112,10 @@ impl ZebraDb {
             // In Zebra 1.4.0 and later, we don't update the sprout tip tree unless it is changed.
             // And we write with a `()` key, not a height key.
             // So we need to look for the most recent update height if the `()` key has never been written.
-            sprout_tree = self.db.zs_last_key_value(&sprout_tree_cf).map(|(_height_key, tree_value)| tree_value);
+            sprout_tree = self
+                .db
+                .zs_last_key_value(&sprout_tree_cf)
+                .map(|(_key: Height, tree_value)| tree_value);
         }
 
         sprout_tree.expect("Sprout note commitment tree must exist if there is a finalized tip")

--- a/zebra-state/src/service/finalized_state/zebra_db/shielded.rs
+++ b/zebra-state/src/service/finalized_state/zebra_db/shielded.rs
@@ -112,7 +112,7 @@ impl ZebraDb {
             // In Zebra 1.4.0 and later, we don't update the sprout tip tree unless it is changed.
             // And we write with a `()` key, not a height key.
             // So we need to look for the most recent update height if the `()` key has never been written.
-            sprout_tree = self.db.zs_last_key_value(&sprout_tree_cf).map(|(_height_key, tree_value)| tree_value);;
+            sprout_tree = self.db.zs_last_key_value(&sprout_tree_cf).map(|(_height_key, tree_value)| tree_value);
         }
 
         sprout_tree.expect("Sprout note commitment tree must exist if there is a finalized tip")

--- a/zebra-state/src/service/finalized_state/zebra_db/shielded.rs
+++ b/zebra-state/src/service/finalized_state/zebra_db/shielded.rs
@@ -112,7 +112,7 @@ impl ZebraDb {
             // In Zebra 1.4.0 and later, we don't update the sprout tip tree unless it is changed.
             // And we write with a `()` key, not a height key.
             // So we need to look for the most recent update height if the `()` key has never been written.
-            sprout_tree = self.db.zs_last_key_value(&sprout_tree_cf);
+            sprout_tree = self.db.zs_last_key_value(&sprout_tree_cf).map(|(_height_key, tree_value)| tree_value);;
         }
 
         sprout_tree.expect("Sprout note commitment tree must exist if there is a finalized tip")

--- a/zebra-state/src/service/finalized_state/zebra_db/shielded.rs
+++ b/zebra-state/src/service/finalized_state/zebra_db/shielded.rs
@@ -109,11 +109,10 @@ impl ZebraDb {
             self.db.zs_get(&sprout_tree_cf, &());
 
         if sprout_tree.is_none() {
-            let tip_height = self
-                .finalized_tip_height()
-                .expect("just checked for an empty database");
-
-            sprout_tree = self.db.zs_get(&sprout_tree_cf, &tip_height);
+            // In Zebra 1.4.0 and later, we don't update the sprout tip tree unless it is changed.
+            // And we write with a `()` key, not a height key.
+            // So we need to look for the most recent update height if the `()` key has never been written.
+            sprout_tree = self.db.zs_last_key_value(&sprout_tree_cf);
         }
 
         sprout_tree.expect("Sprout note commitment tree must exist if there is a finalized tip")

--- a/zebra-state/src/service/finalized_state/zebra_db/shielded.rs
+++ b/zebra-state/src/service/finalized_state/zebra_db/shielded.rs
@@ -115,7 +115,7 @@ impl ZebraDb {
             sprout_tree = self
                 .db
                 .zs_last_key_value(&sprout_tree_cf)
-                .map(|(_key: Height, tree_value)| tree_value);
+                .map(|(_key, tree_value): (Height, _)| tree_value);
         }
 
         sprout_tree.expect("Sprout note commitment tree must exist if there is a finalized tip")

